crates/iota-names-indexer/src/main.rs

```diff
@@ -4,57 +4,162 @@
+    author,
+    version = VERSION,
+    propagate_version = true,
+)]
+enum CLI {
+    Start {
+        /// The URL of an IOTA node to get data from.
+        #[arg(long, default_value = "http://localhost:9000")]
+        node_url: String,
+    },
+}
+
+impl CLI {
+    async fn execute(self) -> Result<()> {
+        match self {
+            CLI::Start { node_url } => {
+                info!("Starting IOTA Names Indexer");
+
+                let prometheus = PrometheusServer::new();
+                let registry = prometheus.registry();
+
+                let cancel_token = CancellationToken::new();
+
+                let mut tasks: JoinSet<Result<()>> = JoinSet::new();
+
+                // Spawn the prometheus API
+                let handle = cancel_token.clone();
+                tasks.spawn(async move {
+                    prometheus.start(handle).await?;
+                    Ok(())
+                });
+
+                // Spawn the metrics worker
+                let handle = cancel_token.clone();
+                tasks.spawn(async move {
+                    let worker = IotaNamesWorker::new(
+                        IotaNamesConfig::from_env().unwrap_or_default(),
+                        Arc::new(IotaNamesMetrics::new(&registry)),
+                    );
+
+                    tokio::select! {
+                        res = worker.run(&node_url, &registry, handle.clone()) => {res?;},
+                        _ = handle.cancelled() => {},
+                    }
+                    Ok(())
+                });
+
+                let mut exit_code = Ok(());
+
+                tokio::select! {
+                    res = interrupt_or_terminate() => {
+                        if let Err(err) = res {
+                            tracing::error!("subscribing to OS interrupt signals failed with error: {err}; shutting down");
+                            exit_code = Err(err);
+                        } else {
+                            tracing::info!("received ctrl-c or terminate; shutting down");
+                        }
+                    },
+                    res = tasks.join_next() => {
+                        if let Some(Ok(Err(err))) = res {
+                            tracing::error!("a worker failed with error: {err}");
+                            exit_code = Err(err);
+                        }
+                    },
+                }
+
+                cancel_token.cancel();
+
+                // Allow the user to abort if the tasks aren't shutting down quickly.
+                tokio::select! {
+                    res = interrupt_or_terminate() => {
+                        if let Err(err) = res {
+                            tracing::error!("subscribing to OS interrupt signals failed with error: {err}; aborting");
+                            exit_code = Err(err);
+                        } else {
+                            tracing::info!("received second ctrl-c or terminate; aborting");
```

Comment:
Why not SIGINT instead of ctrl-c.
Maybe that signal wasn't send via "ctrl-c" in the first place :)